src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray`1.cs

```diff
@@ -875,23 +886,25 @@ public ImmutableArray<T> RemoveAll(Predicate<T> match)
             Requires.NotNull(match, "match");
 
             if (self.IsEmpty)
-            {
-                return new ImmutableArray<T>(self.array);
-            }
+                return self;
 
             List<int> removeIndexes = null;
-            for (int i = 0; i < self.array.Length; i++)
+            
+            int i = 0;
+            for (; i < self.array.Length; i++)
             {
                 if (match(self.array[i]))
                 {
-                    if (removeIndexes == null)
-                    {
-                        removeIndexes = new List<int>();
-                    }
-
+                    removeIndexes = new List<int>();
                     removeIndexes.Add(i);
+                    break;
                 }
             }
+            for (; i < self.array.Length; i++)
+            {
+                if (match(self.array[i]))
+                    removeIndexes.Add(i);
+            }
```

Comment:
Could you share your performance test that demonstrates this change helps?  I understand it in theory.  In practice it makes the code more complicated and duplicates code, so there should be a measurable win to justify it.